### PR TITLE
perf(MuxPath): prefix and fullpath support rewrite_target

### PR DESCRIPTION
**Pull request by aniaan**
close: #552

### CODE REVIEW

File: pkg/object/httpserver/mux.go [hunk 1]
```diff
@@ -261,6 +261,31 @@ func (mp *MuxPath) matchPath(ctx context.HTTPContext) bool {
...
+		path = strings.Replace(path, mp.pathPrefix, mp.rewriteTarget, 1)
+		ctx.Request().SetPath(path)
+		return
+	}
+
+	if mp.pathRE != nil {
+		path = mp.pathRE.ReplaceAllString(path, mp.rewriteTarget)
+		ctx.Request().SetPath(path)
+		return
+	}
```
Comment (localvar) [inline pkg/object/httpserver/mux.go:286]: This is incorrect, for example:
if both `mp.path` and `mp.pathPrefix` are not empty and the request matches `mp.pathPrefix`, the path should be rewritten at L277, but was actually rewritten at L272, which is wrong.

**Comment by samutamm** [comment]
This change could be documented here: [controllers.md](https://github.com/megaease/easegress/blob/main/doc/reference/controllers.md?plain=1#L394) . Old doc

`Use pathRegexp.[ReplaceAllString](https://golang.org/pkg/regexp/#Regexp.ReplaceAllString)(path, rewriteTarget) to rewrite request path` by

New doc:

`Use pathRegexp.[ReplaceAllString](https://golang.org/pkg/regexp/#Regexp.ReplaceAllString)(path, rewriteTarget) or pathPrefix [strings.Replace](https://pkg.go.dev/strings#Replace) to rewrite request path`

File: pkg/object/httpserver/mux.go (hunk 1 above)
Comment (aniaan) [reply pkg/object/httpserver/mux.go:286]: I missed the point that path and prefix and regexp can co-exist, I previously thought they were a mutually exclusive relationship, I'll fix that

**Comment by aniaan** [comment]
1. If path, prefix, and regexp are empty, but rewriteTarget is not empty, the newly created spec will fail to verify. For the existing spec, rewriteTarget will not take effect, because I didn't expect the three to be empty at present, but The scenario where the path is still to be rewritten, so a check is added
2. There is also the handleRewrite function, you can see if it is appropriate to do so

@samutamm @localvar

File: pkg/object/httpserver/spec.go [hunk 2]
```diff
@@ -198,3 +199,13 @@ func (h *Header) Validate() error {
 
 	return nil
 }
+
+// Validate validates Path.
+
```
Comment (samutamm) [inline pkg/object/httpserver/spec.go]: nit: empty line between comment and function

File: pkg/object/httpserver/mux.go [hunk 3]
```diff
@@ -78,18 +78,28 @@ type (
...
 
 	// SearchResult is returned by SearchPath
 	SearchResult string
+
+	// PathType is the type of path
+	PathType string
+)
+
+const (
+	PATH   PathType = "path"
```
Comment (samutamm) [inline pkg/object/httpserver/mux.go:101]: nit: please comment PATH, PREFIX and REGEXP

**Review by samutamm — APPROVED** [review]

File: pkg/object/httpserver/mux.go (hunk 3 above)
Comment (localvar) [reply pkg/object/httpserver/mux.go:101]: I propose not to introduce the `PathType`, it brings extra complexity and only a few benefits.
The `handleRewrite` function could be changed to below and other code could be changes accordingly:
```go
func (mp *MuxPath) handleRewrite(req context.HTTPRequest) {
	if mp.rewriteTarget == "" {
		return
	}

	path := req.Path()

	if mp.path != "" && mp.path == path {
		req.SetPath(mp.rewriteTarget)
		return
	}
	
	if mp.pathPrefix != "" && strings.HasPrefix(path, mp.pathPrefix) {
		path = mp.pathPrefix + path[len(mp.pathPrefix):]
		req.SetPath(path)
		return
	}
	
	// If we get here, then we are sure the following is true:
	//    mp.pathRE != nil && mp.pathRE.MatchString(path)
	path = mp.pathRE.ReplaceAllString(path, mp.rewriteTarget)
	req.SetPath(path)
}
```

**Review by samutamm — APPROVED** [review]

**Pull request merged**